docs/discounts-lists/index.md

```diff
@@ -0,0 +1,496 @@
+  [List.sort](https://melange.re/v3.0.0/api/re/melange/Stdlib/List/#val-sort).
+  `List.sort` returns a brand new list, because, unlike
+`Js.Array.sortInPlaceWith`, it doesn’t modify its argument (and it can’t, since
+  lists are immutable).
+
+## Pattern match on lists
+
+The switch expression in `Discount.getFreeBurgers` accepts the entire `prices`
+list. Unlike with arrays, we can pattern match on lists even if we don’t know
+the length of the list.
+
+Inside the "failure" branch of the switch expression, we see this:
+
+```reason
+| []
+| [_] => None
+```
+
+This pattern matches both empty lists and lists with one element, and returns
+`None`.
+
+Inside the "success" branch of the switch expression, we have:
+
+```reason
+| [_, cheaperPrice, ..._] => Some(cheaperPrice)
+```
+
+This pattern will match on lists that have at least two elements. The first
+element is ignored via `_`. The second element is bound to the name
+`cheaperPrice`, which is encased in `Some` and returned. We use *list spread
+syntax* (`...`) to indicate that the list can have more than the two elements we
+explicitly matched on.
+
+## List spread syntax
+
+The spread operator (`...`) is also used to add elements to the front of a list:
+
+```reason
+let list = [1, 2, 3];
+Js.log([0, ...list]); // [0, 1, 2, 3]
+Js.log([-1, 0, ...list]); // [-1, 0, 1, 2, 3]
+```
+
+::: tip
+
+```reason
+[1, 2, 3]
+```
+
+is really just a shortcut for
+
+```reason
+[1, ...[2, ...[3, ...[]]]]
+```
+
+:::
+
+When pattern matching, the spread operator allows you to bind the **tail** of
+a list to a name:
+
+```reason
+switch ([1, 2, 3]) {
+| [_, ...tail] => Js.log(tail) // [2, 3]
+| _ => ()
+};
+
+switch ([1, 2, 3]) {
+| [_, _, _, ...tail] => Js.log(tail) // []
+| _ => ()
+};
+```
+
+The tail of the list is the sublist that remains after you extract the first n
+elements from the front of the list. As you can see, the tail might be the empty
+list (`[]`). In practice, you don't need to bind the tail to a name unless
+you're writing a [custom list function](/todo). Often, you'll just bind the tail
+to `_`, effectively ignoring it:
+
+```reason{2}
+switch (["one", "two", "three"]) {
+| [a, b, ..._] => Js.log(a ++ ", " ++ b ++ ", etc") // one, two, etc
+| _ => ()
+};
+```
+
+## Runtime representation of lists
+
+If you run the code snippets above in the [Melange
+Playground](https://melange.re/v3.0.0/playground/?language=Reason&code=bGV0IGxpc3QgPSBbMSwgMiwgM107CkpzLmxvZyhbMCwgLi4ubGlzdF0pOyAvLyBbMCwgMSwgMiwgM10KSnMubG9nKFstMSwgMCwgLi4ubGlzdF0pOyAvLyBbLTEsIDAsIDEsIDIsIDNdCgpzd2l0Y2ggKFsxLCAyLCAzXSkgewp8IFsxLCAuLi50YWlsXSA9PiBKcy5sb2codGFpbCkgLy8gWzIsIDNdCnwgXyA9PiAoKQp9OwoKc3dpdGNoIChbMSwgMiwgM10pIHsKfCBbMSwgMiwgMywgLi4udGFpbF0gPT4gSnMubG9nKHRhaWwpIC8vIFtdCnwgXyA9PiAoKQp9Owo%3D&live=off),
+you'll see output like this:
+
+```text
+{"hd":0,"tl":{"hd":1,"tl":{"hd":2,"tl":{"hd":3,"tl":0}}}}
+{"hd":-1,"tl":{"hd":0,"tl":{"hd":1,"tl":{"hd":2,"tl":{"hd":3,"tl":0}}}}}
+{"hd":2,"tl":{"hd":3,"tl":0}}
```

Comment:
I wonder if it'd help to show this output formatted, to clarify how it works, e.g.:

```suggestion
you'll see output like this (formatted):

```text
{
  "hd": 0,
  "tl": {
    "hd": 1,
    "tl": {
      "hd": 2,
      "tl": {
        "hd": 3,
        "tl": 0
      }
    }
  }
}
{
  "hd": -1,
  "tl": {
    "hd": 0,
    "tl": {
      "hd": 1,
      "tl": {
        "hd": 2,
        "tl": {
          "hd": 3,
          "tl": 0
        }
      }
    }
  }
}
{
  "hd": 2,
  "tl": {
    "hd": 3,
    "tl": 0
  }
}
```